WalletWasabi/KeyManagement/KeyManager.cs

```diff
@@ -133,21 +144,59 @@ public void SetFilePath(string filePath)
 
 		public void ToFile()
 		{
-			if (FilePath is null) return;
-			ToFile(FilePath);
+			lock (HdPubKeysLock)
+				lock (BlockchainStateLock)
+					lock (ToFileLock)
+					{
+						if (FilePath is null) return;
+						ToFileNoLock(FilePath);
+					}
 		}
 
-		public void ToFile(string filePath)
+		public void ToFileNoBlockchainStateLock()
 		{
-			IoHelpers.EnsureContainingDirectoryExists(filePath);
 			lock (HdPubKeysLock)
-			{
 				lock (ToFileLock)
 				{
-					string jsonString = JsonConvert.SerializeObject(this, Formatting.Indented);
-					File.WriteAllText(filePath, jsonString, Encoding.UTF8);
+					if (FilePath is null) return;
+					ToFileNoLock(FilePath);
 				}
-			}
+		}
+
+		public void ToFile(string filePath)
+		{
+			lock (HdPubKeysLock)
+				lock (BlockchainStateLock)
+					lock (ToFileLock)
+					{
+						ToFileNoLock(filePath);
+					}
+		}
+
+		private void ToFileNoLock()
+		{
+			if (FilePath is null) return;
+			ToFileNoLock(FilePath);
+		}
+
+		private void ToFileNoLock(string filePath)
+		{
+			IoHelpers.EnsureContainingDirectoryExists(filePath);
+			// Remove the last 100 blocks to ensure verification on the next run. This is needed of reorg.
+			var maturity = 101;
+			var prevHeight = BlockchainState.BestHeight.Value;
+			var matureHeight = Math.Max(0, prevHeight - maturity);
+
+			BlockchainState.BestHeight = new Height(matureHeight);
+			var toRemove = BlockchainState.BlockStates.Where(x => x.BlockHeight >= BlockchainState.BestHeight).ToHashSet();
+			BlockchainState.BlockStates.RemoveAll(x => toRemove.Contains(x));
```

Comment:
I think this could be made faster (I didn't test it nor measure neither it but it should be faster); 
```suggestion
			BlockchainState.BestHeight = new Height(matureHeight);
			var tmp = BlockchainState.BlockStates.Where(x => x.BlockHeight < BlockchainState.BestHeight).ToList();
			BlockchainState.BlockStates.RemoveAll(x => x.BlockHeight >= BlockchainState.BestHeight);
```

Reply:
These commit suggestions are neat! But I'm not sure I understand your logic here:

1. You create a list `tmp` with all the non-reorgable block states.
2. You remove all the reorgable block states.

- Why create tmp?

- `RemoveAll` removes the reorgable block states completely. How do you add it back a few lines later if you don't remember it?

Reply:
I simplified the height logic a bit though.

Reply:
I shouldn't have used the github suggestions here because my code is an idea and not a commit suggestion.
My idea is avoid creating a `HastSet` to delete those contained on it. There rest is the same, you have to do `BlockchainState.BlockStates.AddRange(tmp);`.
It is not an important improvement anyway.

Reply:
But that's exactly what we're doing. Anyway I merge if I misunderstood something feel free to make a mini-PR, I like them.